### PR TITLE
Fix leaks in openssl/integer.hpp and refactoring file

### PR DESCRIPTION
**Pull request by Efrat1**
(1) fixed leak in operator=
(2) fixed leak in toHexString()
(3) convert BIGNUM num_ to unique_ptr

* **Problem Overview**  
  Two leaks and raw pointer usage.
* **Testing Done**  
  Apollo and ASAN

### CODE REVIEW

File: util/include/crypto/openssl/integer.hpp [hunk 1]
```diff
@@ -24,121 +24,121 @@
...
+    UniqueBNCTX ctx(BN_CTX_new());
+    assertResultValid(ctx.get());
+    Integer result;
+    assertResultValid(BN_mod(result.num_.get(), num_.get(), i.num_.get(), ctx.get()));
+    return result;
   }
   Integer operator=(const Integer& i) {
-    num_ = BN_dup(i.num_);
-    if (!num_) THROW_OPENSSL_ERROR;
+    num_.reset(BN_dup(i.num_.get()));
```
Comment (WildFireFlum) [inline util/include/crypto/openssl/integer.hpp:117]: need to check for errors, dup can fail

File: util/include/crypto/openssl/integer.hpp [hunk 2]
```diff
@@ -24,121 +24,121 @@
...
+    if (!BN_lshift(result.num_.get(), num_.get(), n)) THROW_OPENSSL_ERROR;
+    return result;
   }
-  size_t size() const { return BN_num_bytes(num_); }
+
+  size_t size() const { return BN_num_bytes(num_.get()); }
   std::string toHexString(bool add_prefix = false) const {
-    if (add_prefix) return std::string("0x") + BN_bn2hex(num_);
-    return BN_bn2hex(num_);
+    std::string result{UniqueOpenSSLString{BN_bn2hex(num_.get())}.get()};
```
Comment (WildFireFlum) [inline util/include/crypto/openssl/integer.hpp:132]: check result for errors

File: util/include/crypto/openssl/integer.hpp [hunk 3]
```diff
@@ -24,121 +24,121 @@
...
+    if (add_prefix) {
+      result = std::string("0x") + result;
+    }
+    return result;
   }
-  std::string toDecString() const { return BN_bn2dec(num_); }
+  std::string toDecString() const { return BN_bn2dec(num_.get()); }
```
Comment (WildFireFlum) [inline util/include/crypto/openssl/integer.hpp:138]: As Toly said, this implementation allocates a string and leaks it (should be protected with UniqueOpenSSLString and checked for errors)

File: util/include/crypto/openssl/crypto.hpp [hunk 4]
```diff
@@ -40,6 +40,8 @@
 
 namespace concord::crypto::openssl {
 
+void _openssl_free(void* ptr);
```
Comment (WildFireFlum) [inline util/include/crypto/openssl/crypto.hpp:43]: Document why the function was declared when the OPENSSL_FREE macro already exists.

**Review by WildFireFlum — CHANGES_REQUESTED** [review]

File: util/include/crypto/openssl/integer.hpp [hunk 5]
```diff
@@ -24,121 +24,121 @@
...
+  bool operator!=(const Integer& i) const { return BN_cmp(num_.get(), i.num_.get()) != 0; }
+  bool operator==(const Integer& i) const { return BN_cmp(num_.get(), i.num_.get()) == 0; }
 
   Integer operator<<(size_t n) const {
-    BIGNUM* res = BN_new();
-    if (!res) THROW_OPENSSL_ERROR;
-    if (!BN_lshift(res, num_, n)) THROW_OPENSSL_ERROR;
-    return res;
+    Integer result;
+    if (!BN_lshift(result.num_.get(), num_.get(), n)) THROW_OPENSSL_ERROR;
```
Comment (glevkovich) [inline util/include/crypto/openssl/integer.hpp:126]: nit: can use the assert macro?

**Review by glevkovich — CHANGES_REQUESTED** [review]

File: util/include/crypto/openssl/integer.hpp [hunk 6]
```diff
@@ -24,121 +24,121 @@
...
-    if (!num_) THROW_OPENSSL_ERROR;
+    assertResultValid(num_.get());
   }
+  ~Integer() {}
+  Integer(const Integer& i) : num_{BN_dup(i.num_.get())} { assertResultValid(num_.get()); }
+  Integer(Integer&& other) { num_ = std::move(other.num_); }
 
- private:
-  Integer(BIGNUM* num) : num_{num} {}
+  template <typename T>
```
[the remark this answers is not in the document]
Comment (WildFireFlum) [reply util/include/crypto/openssl/integer.hpp:52]: Macros are usually used when there is a need to maintain the call site information (for example, when the line is documented). This was not the case in the original version.

File: util/include/crypto/openssl/integer.hpp (hunk 1 above)
Comment (Efrat1) [reply util/include/crypto/openssl/integer.hpp:116]: done

File: util/include/crypto/openssl/integer.hpp (hunk 3 above)
Comment (Efrat1) [reply util/include/crypto/openssl/integer.hpp]: fixes, thanks

File: util/include/crypto/openssl/integer.hpp (hunk 2 above)
Comment (Efrat1) [reply util/include/crypto/openssl/integer.hpp]: done

File: util/include/crypto/openssl/crypto.hpp (hunk 4 above)
Comment (Efrat1) [reply util/include/crypto/openssl/crypto.hpp:44]: added

File: util/include/crypto/openssl/integer.hpp [hunk 7]
```diff
@@ -24,121 +24,121 @@
...
     if (val < 0) {
-      BN_set_negative(num_, 1);
+      BN_set_negative(num_.get(), 1);
       val *= -1;
     }
-    if (!BN_set_word(num_, val)) THROW_OPENSSL_ERROR;
+    assertResultValid(BN_set_word(num_.get(), val));
   }
-  Integer(const std::string& s) : Integer{reinterpret_cast<unsigned const char*>(s.data()), s.size()} {}
+  Integer(const std::string& s) : Integer(reinterpret_cast<unsigned const char*>(s.data()), s.size()) {}
```
[the remark this answers is not in the document]
Comment (Efrat1) [reply util/include/crypto/openssl/integer.hpp]: thanks

File: util/include/crypto/openssl/integer.hpp [hunk 8]
```diff
@@ -24,121 +24,121 @@
...
   // BN_bin2bn() converts the positive integer in big-endian form of length len at s into a BIGNUM
   Integer(const unsigned char* val_ptr, size_t size) : num_{BN_bin2bn(val_ptr, size, nullptr)} {
-    if (!num_) THROW_OPENSSL_ERROR;
-  }
-  ~Integer() { BN_free(num_); }
-  Integer(const Integer& i) : num_{BN_dup(i.num_)} {
-    if (!num_) THROW_OPENSSL_ERROR;
+    assertResultValid(num_.get());
   }
+  ~Integer() {}
```
[the remark this answers is not in the document]
Comment (Efrat1) [reply util/include/crypto/openssl/integer.hpp]: done

File: util/include/crypto/openssl/integer.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (Efrat1) [reply util/include/crypto/openssl/integer.hpp:52]: I kept it as is

File: util/include/crypto/openssl/integer.hpp (hunk 5 above)
Comment (Efrat1) [reply util/include/crypto/openssl/integer.hpp]: sure, missed that, thanks.

File: util/include/crypto/openssl/integer.hpp [hunk 9]
```diff
@@ -24,121 +24,121 @@
...
+    return result;
+  }
+  void setNegative() { BN_set_negative(num_.get(), 1); }
+  bool isNegative() { return BN_is_negative(num_.get()); }
 
   Integer operator+(const Integer& i) {
-    BIGNUM* res = BN_new();
-    BN_add(res, num_, i.num_);
-    return res;
+    Integer result;
```
[the remark this answers is not in the document]
Comment (Efrat1) [reply util/include/crypto/openssl/integer.hpp]: thanks, done.

**Review by WildFireFlum — APPROVED** [review]

**Comment by Efrat1** [comment]
Push these changes to v0.16.x and glevkovich/memory_management_dev1 which will be merged to master in future.
closing this PR.

**Comment by glevkovich** [comment]
@Efrat1 
To make things clear, we agreed that this PR is going to be closed, and this branch is merged (after PR) into glevkovich/memory_management_dev1.

**Pull request closed without merging**